File: src/lib/ai-ranking.ts
import type { HospitalWithScore } from "@/types/hospital";

const ANTHROPIC_API_KEY = process.env.ANTHROPIC_API_KEY || "";

interface AIRankingResult {
  hospitals: HospitalWithScore[];
}

export async function getAIRanking(
  hospitals: HospitalWithScore[],
  severity: string = "medium"
): Promise<AIRankingResult> {
  if (!ANTHROPIC_API_KEY) {
    // Return hospitals as-is if no API key configured
    return { hospitals };
  }

  const hospitalSummary = hospitals
    .map(
      (h) =>
        `- ${h.name}: Wait ${h.waitTimeLabel}, Travel ${h.travelTimeMinutes}min (${h.distanceKm}km), Score ${h.priorityScore}`
    )
    .join("\n");

  const prompt = `You are a medical triage assistant helping recommend hospitals.

Patient severity: ${severity}

Available hospitals (already scored by wait time + travel time):
${hospitalSummary}

For each hospital, provide a brief 1-sentence reasoning for its ranking.
Consider: wait times, travel distance, hospital specialization, and severity level.

Respond in JSON format:
{
  "rankings": [
    { "name": "Hospital Name", "reasoning": "Brief explanation" }
  ]
}`;

  try {
    const response = await fetch("https://api.anthropic.com/v1/messages", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "x-api-key": ANTHROPIC_API_KEY,
        "anthropic-version": "2023-06-01",
      },
      body: JSON.stringify({
        model: "claude-sonnet-4-20250514",
        max_tokens: 1024,
        messages: [{ role: "user", content: prompt }],
      }),
    });

    if (!response.ok) {
      console.error("Anthropic API error:", response.statusText);
      return { hospitals };
    }

    const data = await response.json();
    const text =
      data.content?.[0]?.text || "";

    // Parse the JSON from the response
    const jsonMatch = text.match(/\{[\s\S]*\}/);
    if (jsonMatch) {
      const parsed = JSON.parse(jsonMatch[0]);
      const rankings = parsed.rankings || [];

      // Merge AI reasoning into hospital data
      const enriched = hospitals.map((h) => {
        const aiEntry = rankings.find(
          (r: { name: string; reasoning: string }) =>
            r.name.toLowerCase().includes(h.name.toLowerCase()) ||
            h.name.toLowerCase().includes(r.name.toLowerCase())
        );
        return {
          ...h,
          aiReasoning: aiEntry?.reasoning || undefined,
        };
      });

      return { hospitals: enriched };
    }
  } catch (error) {
    console.error("AI ranking error:", error);
  }

  return { hospitals };
}

File: backend/app/services/ai_ranking.py
import json

import anthropic

from app.config import get_settings
from app.models import HospitalWithScore


async def get_ai_ranking(
    hospitals: list[HospitalWithScore],
    severity: str = "medium",
) -> list[HospitalWithScore]:
    """Enrich hospital rankings with AI-generated reasoning."""
    settings = get_settings()

    if not settings.anthropic_api_key:
        return hospitals

    summary = "\n".join(
        f"- {h.name}: Wait {h.wait_time_label}, "
        f"Travel {h.travel_time_minutes}min ({h.distance_km}km), "
        f"Score {h.priority_score}"
        for h in hospitals
    )

    prompt = (
        "You are a medical triage assistant helping recommend hospitals.\n\n"
        f"Patient severity: {severity}\n\n"
        f"Available hospitals (already scored by wait time + travel time):\n{summary}\n\n"
        "For each hospital, provide a brief 1-sentence reasoning for its ranking.\n"
        "Consider: wait times, travel distance, hospital specialization, and severity level.\n\n"
        'Respond in JSON format:\n'
        '{\n  "rankings": [\n'
        '    { "name": "Hospital Name", "reasoning": "Brief explanation" }\n'
        "  ]\n}"
    )

    try:
        client = anthropic.AsyncAnthropic(api_key=settings.anthropic_api_key)
        message = await client.messages.create(
            model="claude-sonnet-4-20250514",
            max_tokens=1024,
            messages=[{"role": "user", "content": prompt}],
        )

        text = message.content[0].text if message.content else ""

        # Extract JSON from response
        start = text.find("{")
        end = text.rfind("}") + 1
        if start >= 0 and end > start:
            parsed = json.loads(text[start:end])
            rankings = parsed.get("rankings", [])

            for h in hospitals:
                match = next(
                    (
                        r
                        for r in rankings
                        if r["name"].lower() in h.name.lower()
                        or h.name.lower() in r["name"].lower()
                    ),
                    None,
                )
                if match:
                    h.ai_reasoning = match.get("reasoning")

    except Exception as e:
        print(f"AI ranking error: {e}")

    return hospitals
